### PR TITLE
Fix duplicating diagnoses when saving a visit note and implements deleting of diagnoses on a visit note.

### PR DESCRIPTION
**Pull request by pluwum**
The PR fixes duplicating diagnoses when saving a visit note and implements deleting of diagnoses on a visit note.

## Issue I worked on
see https://issues.openmrs.org/browse/TRUNK-5343

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

- [ ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

- [ ] All new and existing **tests passed**.

- [x] My pull request is **based on the latest changes** of the master branch.

### CODE REVIEW

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java [hunk 1]
```diff
@@ -121,7 +122,6 @@ public String generateHtml(FormEntryContext context) {
             try {
                 Map<String, Object> fragmentConfig = new HashMap<String, Object>();
                 fragmentConfig.put("formFieldName", "encounterDiagnoses");
-                fragmentConfig.put("existingDiagnoses", existingDiagnoses);
```
Comment (dkayiwa) [inline api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: Why did you remove the above?

**Comment by wluyima** [comment]
It appears to me like the ticket description and this PR are not in sync

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java (hunk 1 above)
Comment (pluwum) [reply api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: @dkayiwa i had removed i realised i had removed it while reviewing and i put it back

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java [hunk 2]
```diff
@@ -220,11 +230,18 @@ public void handleSubmission(FormEntrySession formEntrySession, HttpServletReque
                     System.out.println(nd);
                 }
 
+                if (node.path("existingDiagnosis").getNumberValue() != null) {
+                    JsonNode nd = node.get("existingDiagnosis");
+                    System.out.println(nd);
```
Comment (dkayiwa) [inline api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: Do you need the above println?

File: omod/src/main/java/org/openmrs/module/coreapps/fragment/controller/diagnosis/EncounterDiagnosesFragmentController.java [hunk 3]
```diff
@@ -54,7 +54,7 @@ public int compare(Diagnosis left, Diagnosis right) {
                 SimpleObject simple = diagnosesFragmentController.simplify(csr, ui, Context.getLocale());
                 jsDiagnosis = simple.toJson();
             }
-            jsForDiagnoses.add("{ diagnosis: diagnoses.CodedOrFreeTextConceptAnswer(" + jsDiagnosis + "), confirmed: " + (d.getCertainty().equals(Diagnosis.Certainty.CONFIRMED)) + ", primary: " + (d.getOrder().equals(Diagnosis.Order.PRIMARY)) + ", existingObs: " + (d.getExistingObs() != null ? d.getExistingObs().getId() : null) + " }");
+            jsForDiagnoses.add("{ diagnosis: diagnoses.CodedOrFreeTextConceptAnswer(" + jsDiagnosis + "), confirmed: " + (d.getCertainty().equals(Diagnosis.Certainty.CONFIRMED)) + ", primary: " + (d.getOrder().equals(Diagnosis.Order.PRIMARY)) + ", existingObs: " + (d.getExistingObs() != null ? d.getExistingObs().getId() : null) + ", existingDiagnosis: " + d.getExistingDiagnosis() + " }");
```
Comment (dkayiwa) [inline omod/src/main/java/org/openmrs/module/coreapps/fragment/controller/diagnosis/EncounterDiagnosesFragmentController.java]: Did you test this for cases where existing diagnosis returns null?

**Comment by dkayiwa** [comment]
The commit message and pull request title do not seem to reflect the purpose of the commit.

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java [hunk 4]
```diff
@@ -209,26 +216,44 @@ public void handleSubmission(FormEntrySession formEntrySession, HttpServletReque
...
-                diagnosis.setPatient(formEntrySession.getPatient());
-                diagnosis.setDiagnosis(new CodedOrFreeText(answer.getCodedAnswer(), answer.getSpecificCodedAnswer(), answer.getNonCodedAnswer()));
+                if (node.path("existingDiagnosis").getNumberValue() != null) {
+                    existingDiagnosis = node.get("existingDiagnosis").getIntValue();
+                }
+
+                org.openmrs.Diagnosis diagnosis;
+
+                if(existingDiagnosis !=null){
+                    diagnosis = Context.getDiagnosisService().getDiagnosis(existingDiagnosis);
```
Comment (dkayiwa) [inline api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: Do you need to make the getDiagnosis call twice?

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java [hunk 5]
```diff
@@ -209,26 +216,44 @@ public void handleSubmission(FormEntrySession formEntrySession, HttpServletReque
...
+                    diagnosis.setDiagnosis(Context.getDiagnosisService().getDiagnosis(existingDiagnosis).getDiagnosis());
+                    diagnosis.setDateChanged(new Date());
```
Comment (dkayiwa) [inline api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: Can you save an edited diagnosis only if any of its properties has been changed?

File: omod/src/main/java/org/openmrs/module/coreapps/fragment/controller/diagnosis/EncounterDiagnosesFragmentController.java (hunk 3 above)
Comment (pluwum) [reply omod/src/main/java/org/openmrs/module/coreapps/fragment/controller/diagnosis/EncounterDiagnosesFragmentController.java]: @dkayiwa is [this](https://github.com/pluwum/openmrs-module-coreapps/blob/aa34c02b443e4b63acf55966e15d1511a727edc0/omod/src/test/java/org/openmrs/module/coreapps/fragment/controller/diagnosis/EncounterDiagnosesFragmentControllerTest.java#L60) what you mean?

Comment (dkayiwa) [reply omod/src/main/java/org/openmrs/module/coreapps/fragment/controller/diagnosis/EncounterDiagnosesFragmentController.java]: Looks ok.

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java [hunk 6]
```diff
@@ -209,27 +222,74 @@ public void handleSubmission(FormEntrySession formEntrySession, HttpServletReque
...
+                        diagnosis.setChangedBy(Context.getAuthenticatedUser());
+                        Context.getDiagnosisService().save(diagnosis);
+                    }
+
+                }else{
+                    diagnosis = new org.openmrs.Diagnosis();
+                    diagnosis.setDiagnosis(new CodedOrFreeText(answer.getCodedAnswer(), answer.getSpecificCodedAnswer(), answer.getNonCodedAnswer()));
+                    diagnosis.setEncounter(formEntrySession.getEncounter());
+                    diagnosis.setCertainty(certaintyStatus);
+                    diagnosis.setRank(orderId);
```
Comment (dkayiwa) [inline api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: How does order id become rank?

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java [hunk 7]
```diff
@@ -209,27 +222,74 @@ public void handleSubmission(FormEntrySession formEntrySession, HttpServletReque
...
+                    diagnosis.setPatient(formEntrySession.getPatient());
+                    diagnosis.getEncounter().setEncounterType(Context.getEncounterService().getEncounterTypeByUuid("d7151f82-c1f3-4152-a605-2f9ea7414a79"));
```
Comment (dkayiwa) [inline api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: Can we use a constant for the uuid? Something like VISIT_NOTE_ENCOUNTER_TYPE_UUID

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java [hunk 8]
```diff
@@ -209,27 +222,74 @@ public void handleSubmission(FormEntrySession formEntrySession, HttpServletReque
...
+
+                    // Check if an existing diagnosis is being resubmitted
+                    if (!existingDiagnoses.add(diagnosis)) {
+                        resubmittedDiagnoses.add(diagnosis);
+                    }
+
+                    if(!diagnosis.getRank().equals(orderId) || !diagnosis.getCertainty().equals(certaintyStatus)){
+                        diagnosis.setRank(orderId);
+                        diagnosis.setCertainty(certaintyStatus);
+                        diagnosis.setDiagnosis(diagnosis.getDiagnosis());
```
Comment (dkayiwa) [inline api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: What is the above line doing?

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java [hunk 9]
```diff
@@ -209,27 +222,74 @@ public void handleSubmission(FormEntrySession formEntrySession, HttpServletReque
...
-                diagnosis.setRank(diagnosisOrder == Diagnosis.Order.PRIMARY ? 1 : 2);
-                diagnosis.setPatient(formEntrySession.getPatient());
-                diagnosis.setDiagnosis(new CodedOrFreeText(answer.getCodedAnswer(), answer.getSpecificCodedAnswer(), answer.getNonCodedAnswer()));
+                if (node.path("existingDiagnosis").getNumberValue() != null) {
+                    existingDiagnosis = node.get("existingDiagnosis").getIntValue();
+                }
+
+                org.openmrs.Diagnosis diagnosis;
+
+                Integer orderId = diagnosisOrder == Diagnosis.Order.PRIMARY ? 1 : 2;
```
Comment (dkayiwa) [inline api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: Can we use constant variables for 1 and 2

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java [hunk 10]
```diff
@@ -209,27 +222,74 @@ public void handleSubmission(FormEntrySession formEntrySession, HttpServletReque
...
+                if(existingDiagnosis !=null){
+
+                    diagnosis = Context.getDiagnosisService().getDiagnosis(existingDiagnosis);
+
+                    // Check if an existing diagnosis is being resubmitted
+                    if (!existingDiagnoses.add(diagnosis)) {
+                        resubmittedDiagnoses.add(diagnosis);
+                    }
+
+                    if(!diagnosis.getRank().equals(orderId) || !diagnosis.getCertainty().equals(certaintyStatus)){
```
Comment (dkayiwa) [inline api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: Is it only certainty and rank that the UI lets you change?

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java [hunk 11]
```diff
@@ -209,27 +222,74 @@ public void handleSubmission(FormEntrySession formEntrySession, HttpServletReque
...
+                    diagnosis.setPatient(formEntrySession.getPatient());
+                    diagnosis.getEncounter().setEncounterType(Context.getEncounterService().getEncounterTypeByUuid("d7151f82-c1f3-4152-a605-2f9ea7414a79"));
+                    Context.getEncounterService().saveEncounter(diagnosis.getEncounter());
+                    Context.getDiagnosisService().save(diagnosis);
+                }
+
+            }
+
+            if (formEntrySession.getContext().getMode().equals(FormEntryContext.Mode.EDIT)) {
+                // Remove get and deleted Diagnoses that where not resubmitted
```
Comment (dkayiwa) [inline api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: What does "Remove get" mean?

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java [hunk 12]
```diff
@@ -209,27 +222,74 @@ public void handleSubmission(FormEntrySession formEntrySession, HttpServletReque
...
+                Collection<org.openmrs.Diagnosis> diagnosesToVoid = CollectionUtils
+                        .subtract(existingDiagnoses, resubmittedDiagnoses);
+                for (org.openmrs.Diagnosis diagnosisToVoid : diagnosesToVoid) {
+                    diagnosisToVoid.setVoided(true);
```
Comment (dkayiwa) [inline api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: Shouldn't you be using DiagnosisService.voidDiagnosis()?

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java (hunk 10 above)
Comment (pluwum) [reply api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: @dkayiwa Those are all the attributes I can see. am I missing something here?

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java (hunk 8 above)
Comment (pluwum) [reply api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: Setting a diagnosis but since it is an existing diagnosis, i have realized that it is redundant and unnecessary. i have removed it.

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java (hunk 11 above)
Comment (pluwum) [reply api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: Sorry, a grammatical error. It has been corrected.

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java [hunk 13]
```diff
@@ -209,27 +226,72 @@ public void handleSubmission(FormEntrySession formEntrySession, HttpServletReque
...
-                Context.getEncounterService().saveEncounter(diagnosis.getEncounter());
-                Context.getDiagnosisService().save(diagnosis);
+                Integer rank = diagnosisOrder == Diagnosis.Order.PRIMARY ? DIAGNOSIS_RANK_PRIMARY : DIAGNOSIS_RANK_SECONDARY;
+                ConditionVerificationStatus certaintyStatus = certainty == Diagnosis.Certainty.CONFIRMED ? ConditionVerificationStatus.CONFIRMED : ConditionVerificationStatus.PROVISIONAL;
+
+                if(existingDiagnosis !=null){
+
+                    diagnosis = Context.getDiagnosisService().getDiagnosis(existingDiagnosis);
+
+                    // Check if an existing diagnosis is being resubmitted
```
Comment (dkayiwa) [inline api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: Are you checking the existingDiagnoses collection or adding to it? (That is in reference to your comment above)

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java [hunk 14]
```diff
@@ -209,27 +226,72 @@ public void handleSubmission(FormEntrySession formEntrySession, HttpServletReque
...
+                    diagnosis.setPatient(formEntrySession.getPatient());
+                    diagnosis.getEncounter().setEncounterType(Context.getEncounterService().getEncounterTypeByUuid(VISIT_NOTE_ENCOUNTER_TYPE_UUID));
+                    Context.getEncounterService().saveEncounter(diagnosis.getEncounter());
+                    Context.getDiagnosisService().save(diagnosis);
+                }
+
+            }
+
+            if (formEntrySession.getContext().getMode().equals(FormEntryContext.Mode.EDIT)) {
+                // Remove Diagnoses that where not resubmitted
```
Comment (dkayiwa) [inline api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: where or were?

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java (hunk 13 above)
Comment (CeciliaCaroline) [reply api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: Have removed the if statement because it was redundant since no duplicates can be added to a set so no need to check for that.
